api/metrics-api-jaxrs/src/main/java/org/hawkular/metrics/api/jaxrs/GuageHandler.java

```diff
@@ -0,0 +1,407 @@
+
+                    if (bucketsCount == null && bucketDuration == null) {
+                        ListenableFuture<List<GuageData>> dataFuture = metricsService.findNumericData(tenantId,
+                                new MetricId(id), startTime, endTime);
+                        return Futures.transform(dataFuture, ApiUtils.MAP_COLLECTION);
+                    }
+
+                    if (bucketsCount != null && bucketDuration != null) {
+                        return badRequest(new ApiError("Both buckets and bucketDuration parameters are used"));
+                    }
+
+                    Buckets buckets;
+                    try {
+                        if (bucketsCount != null) {
+                            buckets = Buckets.fromCount(startTime, endTime, bucketsCount);
+                        } else {
+                            buckets = Buckets.fromStep(startTime, endTime, bucketDuration.toMillis());
+                        }
+                    } catch (IllegalArgumentException e) {
+                        return badRequest(new ApiError("Bucket: " + e.getMessage()));
+                    }
+
+                    ListenableFuture<BucketedOutput<NumericBucketDataPoint>> dataFuture;
+                    dataFuture = metricsService.findNumericStats(metric, startTime, endTime, buckets);
+
+                    ListenableFuture<List<NumericBucketDataPoint>> outputFuture;
+                    outputFuture = Futures.transform(dataFuture, BucketedOutput<NumericBucketDataPoint>::getData);
+
+                    return Futures.transform(outputFuture, ApiUtils.MAP_COLLECTION);
+                });
+    }
+
+    @GET
+    @Path("/{tenantId}/guage/{id}/periods")
+    @ApiOperation(value = "Retrieve periods for which the condition holds true for each consecutive data point.",
+        response = List.class)
+    @ApiResponses(value = { @ApiResponse(code = 200, message = "Successfully fetched periods."),
+            @ApiResponse(code = 204, message = "No data was found."),
+            @ApiResponse(code = 400, message = "Missing or invalid query parameters") })
+    public void findPeriods(
+            @Suspended final AsyncResponse asyncResponse,
+            @PathParam("tenantId") String tenantId,
+            @PathParam("id") String id,
+            @ApiParam(value = "Defaults to now - 8 hours", required = false) @QueryParam("start") final Long start,
+            @ApiParam(value = "Defaults to now", required = false) @QueryParam("end") final Long end,
+            @ApiParam(value = "A threshold against which values are compared", required = true)
+                @QueryParam("threshold") double threshold,
+            @ApiParam(value = "A comparison operation to perform between values and the threshold."
+                    + " Supported operations include ge, gte, lt, lte, and eq", required = true)
+                @QueryParam("op") String operator) {
+
+        executeAsync(
```

Comment:
Very long anonymous function here also, and with some business logic also instead of only interface-related stuff.